fix(ivy): ngtsc throws if "flatModuleOutFile" is set to null

devversion opened the pull request:
In ngc is was valid to set the "flatModuleOutFile" option to "null". This is sometimes
necessary if a tsconfig extends from another one but the "fatModuleOutFile" option
needs to be unset (note that "undefined" does not exist as value in JSON)

Now if ngtsc is used to compile the project, ngtsc will fail with an error because it
tries to do string manipulation on the "flatModuleOutFile". This happens because
ngtsc only skips flat module indices if the option is set to "undefined".

Since this is not compatible with what was supported in ngc and such exceptions
should be avoided, the flat module check is now aligned with ngc.

```
TypeError: Cannot read property 'replace' of null
    at Object.normalizeSeparators (/home/circleci/project/node_modules/@angular/compiler-cli/src/ngtsc/util/src/path.js:35:21)
    at new NgtscProgram (/home/circleci/project/node_modules/@angular/compiler-cli/src/ngtsc/program.js:126:52)
```

Additionally setting the `flatModuleOutFile` option to an empty string
currently results in unexpected behavior. No errors is thrown, but the
flat module index file will be `.ts` (no file name; just extension). This is now
also fixed by treating an empty string similarly to `null` and `undefined`.

alfaproject commented on packages/compiler-cli/src/ngtsc/program.ts:
```diff
@@ -116,7 +116,7 @@ export class NgtscProgram implements api.Program {
 
     let entryPoint: AbsoluteFsPath|null = null;
-    if (options.flatModuleOutFile !== undefined) {
+    if (options.flatModuleOutFile) {
```
`!= null` covers both and is more inline with recent code changes

devversion replied:
I had that initially but I wanted to make it behave completely similar to what NGC did (i.e. not throwing if zero is specified, or not doing breaking if an empty string is specified)

alxhub replied:
I agree with @alfaproject actually - let's explicitly check for the conditions we care about. Excluding the empty string does seem like a good idea as well as `null` and `undefined`.

devversion replied:
Sure, I don't feel too strong with it, but it doesn't match NGC's behavior exactly (which I guess is fine for such cases)

devversion replied:
Done. Used `!=` to capture `undefined` and `null` and `!== ''` to capture an empty string. Also added a test for the empty string case.

devversion committed:
fix(ivy): ngtsc throws if "flatModuleOutFile" is set to null

In ngc is was valid to set the "flatModuleOutFile" option to "null". This is sometimes
necessary if a tsconfig extends from another one but the "fatModuleOutFile" option
needs to be unset (note that "undefined" does not exist as value in JSON)

Now if ngtsc is used to compile the project, ngtsc will fail with an error because it
tries to do string manipulation on the "flatModuleOutFile". This happens because
ngtsc only skips flat module indices if the option is set to "undefined".

Since this is not compatible with what was supported in ngc and such exceptions
should be avoided, the flat module check is now aligned with ngc.

```
TypeError: Cannot read property 'replace' of null
    at Object.normalizeSeparators (/home/circleci/project/node_modules/@angular/compiler-cli/src/ngtsc/util/src/path.js:35:21)
    at new NgtscProgram (/home/circleci/project/node_modules/@angular/compiler-cli/src/ngtsc/program.js:126:52)
```

Additionally setting the `flatModuleOutFile` option to an empty string
currently results in unexpected behavior. No errors is thrown, but the
flat module index file will be `.ts` (no file name; just extension).

This is now also fixed by treating an empty string similarly to
`null`.

AndrewKushnir commented:
- [VE Presubmit](http://test/OCL:264698350:BASE:264698371:1566424338498:aa2d3f99)
- [Ivy Presubmit](http://test/OCL:264698350:BASE:264698511:1566424373636:6213b662)
